src/providers/mbed_provider/asym_encryption.rs

```diff
@@ -0,0 +1,94 @@
+        op: psa_asymmetric_encrypt::Operation,
+    ) -> Result<psa_asymmetric_encrypt::Result> {
+        info!("Mbed Provider - Asym Encrypt");
+        let key_name = op.key_name.clone();
+
+        let key_triple = KeyTriple::new(app_name, ProviderID::MbedCrypto, key_name);
+        let store_handle = self.key_info_store.read().expect("Key store lock poisoned");
+        let key_id = key_management::get_key_id(&key_triple, &*store_handle)?;
+        let _guard = self
+            .key_handle_mutex
+            .lock()
+            .expect("Grabbing key handle mutex failed");
+        let id = key::Id::from_persistent_key_id(key_id);
+        let key_attributes = key::Attributes::from_key_id(id)?;
+
+        op.validate(key_attributes)?;
+        let salt_buff = op.salt.as_ref().map(|salt| salt.as_slice());
+        let alg = op.alg;
+        let buffer_size = key_attributes.asymmetric_encrypt_output_size(alg)?;
+        let mut ciphertext = vec![0u8; buffer_size];
+
+        match asym_encryption::encrypt(id, alg, &op.plaintext, salt_buff, &mut ciphertext) {
+            Ok(output_size) => {
+                ciphertext.resize(output_size, 0);
+                Ok(psa_asymmetric_encrypt::Result {
+                    ciphertext: ciphertext.into(),
+                })
+            }
+            Err(error) => {
+                let error = ResponseStatus::from(error);
+                format_error!("Encrypt status: ", error);
+                Err(error)
+            }
+        }
+    }
+
+    pub(super) fn psa_asymmetric_decrypt_internal(
+        &self,
+        app_name: ApplicationName,
+        op: psa_asymmetric_decrypt::Operation,
+    ) -> Result<psa_asymmetric_decrypt::Result> {
+        info!("Mbed Provider - Asym Decrypt");
+
+        let key_triple = KeyTriple::new(app_name, ProviderID::MbedCrypto, op.key_name.clone());
+        let store_handle = self.key_info_store.read().expect("Key store lock poisoned");
+        let key_id = key_management::get_key_id(&key_triple, &*store_handle)?;
+
+        let _guard = self
+            .key_handle_mutex
+            .lock()
+            .expect("Grabbing key handle mutex failed");
+
+        let id = key::Id::from_persistent_key_id(key_id);
+        let key_attributes = key::Attributes::from_key_id(id)?;
+        op.validate(key_attributes)?;
+        let salt_buff = match &op.salt {
+            Some(salt) => Some(salt.as_slice()),
+            None => None,
+        };
+        let buffer_size = key_attributes.asymmetric_decrypt_output_size(op.alg)?;
+        let mut plaintext = vec![0u8; buffer_size];
+
+        match asym_encryption::decrypt(id, op.alg, &op.ciphertext, salt_buff, &mut plaintext) {
+            Ok(output_size) => {
+                plaintext.resize(output_size, 0);
+                Ok(psa_asymmetric_decrypt::Result {
+                    plaintext: plaintext.into(),
+                })
+            }
+            Err(error) => {
+                plaintext.zeroize();
```

Comment:
Ah good point! Do you think, even in case of error, it might contain some decrypted data? Does the spec say anything about it?

Reply:
hmm it doesn't mention it in the spec that I can see. A quick look at the `mbedtls` code shows that it [only copies into plaintext buffer if successful](https://github.com/ARMmbed/mbedtls/blob/527b87890db72da39f6185f977b8230655a42ddd/library/rsa.c#L1712).

Reply:
Yes I reckon that is the safe, default, behaviour that you would expect. We can remove the `zeroize` here I think.